src/script/common/c_content.cpp

```diff
@@ -2009,6 +2010,45 @@ void push_noiseparams(lua_State *L, NoiseParams *np)
 	lua_setfield(L, -2, "spread");
 }
 
+bool read_tree_def(lua_State *L, int idx, const NodeDefManager *ndef,
+		treegen::TreeDef &tree_def)
+{
+	std::string trunk, leaves, fruit;
+	if (!lua_istable(L, idx))
+		return false;
+
+	getstringfield(L, idx, "axiom", tree_def.initial_axiom);
+	getstringfield(L, idx, "rules_a", tree_def.rules_a);
+	getstringfield(L, idx, "rules_b", tree_def.rules_b);
+	getstringfield(L, idx, "rules_c", tree_def.rules_c);
+	getstringfield(L, idx, "rules_d", tree_def.rules_d);
+	getstringfield(L, idx, "trunk", trunk);
+	tree_def.trunknode = ndef->getId(trunk);
```

Comment:
I know this code was not modified but the use-case is slightly different to `minetest.spawn_tree`. Decorations use the NodeResolver to properly deal with `unregister_item` and new aliases.
`TreeDef` would need to be like `Schematic` or `Decoration`: inherit `NodeResolver` and use `m_nodenames` + `m_nnlistsizes`.

Reply:
Thanks for pointing this out, I completely missed it.
Now it uses a `NodeResolverr`, I hope I implemented it the right way.

(Also, I fixed the `place_on` behavior, maybe `canPlaceDecoration` should be called before any decoration function gets called.)